File: HW1/Qsort.scala
//Phuong Pham
//CS 558
//Fall 2018
//Quicksort in Scala

import scala.io.StdIn.{readLine, readInt}
import scala.util.Random

object Constants
{
  val minsize = 10
}

object Qsort
{
  val usage = "\nUsage: Enter a positive integer to generate an array of random numbers.\n"

  //Swap two array elements, a(i) <-> a(j)
  def swap(a: Array[Int], i: Int, j: Int): Unit = 
  {
    if (i == j)
      return
    val temp = a(i)
    a(i) = a(j)
    a(j) = temp
  }

  //Create an array to contain a random permutation of 1..n
  def createArray(n: Int): Array[Int] = 
  {
    Stream.continually(Random.nextInt).take(n).toArray
  }

  //Bubble sort for the base cases
  def bubblesort(a: Array[Int], low: Int, high: Int): Unit = 
  {
    if (low >= high)
      return
    for (i <- low to high)
    {
      for (j <- i + 1 to high)
      {
        if (a(i) > a(j))
          swap(a, i, j)
      }
    }
  }

  //Partition an array range a[low, high] into two sub-ranges
  def partition(a: Array[Int], low: Int, high: Int): Int =
  {
    var pivot = a(high)
    var middle = low

    for (i <- low until high)
    {
      if (a(i) < pivot)
      {
        swap(a, i, middle)
        middle += 1
      }
    }

    swap(a, high, middle)
    return middle
  }

  //Quicksort an array range a(low, high)
  def qsort(a: Array[Int], low: Int, high: Int): Unit = 
  {
    if (high - low < Constants.minsize)
      bubblesort(a, low, high)
    else
    {
      var middle = partition(a, low, high)
      if (low < middle)
        qsort(a, low, middle - 1)
      if (middle < high)
        qsort(a, middle + 1, high)
    }
  }

  //Print an array
  def printArray(a: Array[Int]): Unit = 
  {
    println(a.mkString(", "))
  }

  //Verify a sorted array
  def verifyArray(a: Array[Int], n: Int): Unit = 
  {
    for (i <- 0 until n - 1)
    {
      if (a(i) > a(i + 1))
      {
        println(s"\nFAILED: a($i) = " + a(i) + s" a(${i+1}) = " + a(i+1))
	return
      }
    }
    println("\nResult verified!")
  }

  //Main routine for testing quicksort
  def main(args: Array[String])
  {
    if (args.length != 1)
    {
      println(usage)
      return
    }

    val n = args(0).toInt
    if (n < 1)
    {
      println(usage)
      return
    }
    if (n == 1)
      println("\nSorting a 1 item list? Alright, here it is:")

    println(s"\nGenerating an array of $n random elements...")
    val a = createArray(n)
    println("\nInitial array:")
    printArray(a)

    println("\nSorting array...")
    qsort(a, 0, n - 1)
    
    printf("\nAfter quicksort:\n")
    printArray(a)

    if (n == 1)
      println("\nThat was fun.")

    verifyArray(a, n)
  }
}

File: HW1/qsort.c
//-------------------------------------------------------------------------
// Supporting code for CS558 Programming Languages. 
// Dept of Computer Science, Portland State University
// Original version: J. Li (8/2018)
//-------------------------------------------------------------------------

// Quiksort in C
//
// Usage: linux> ./qsort N
// 
// 
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>

#define MINSIZE   10 

// Swap two array elements, a[i] <-> a[j]
void swap(int *a, int i, int j) {
  if (i == j) return;
  int tmp = a[i];
  a[i] = a[j];
  a[j] = tmp;
}

// Create an array to contain a random permutation of 1..N
int *createArray(int N) {
  int *a = (int *) malloc(sizeof(int) * N);
  int i, j;
  for (i=0; i<N; i++)
    a[i] = i + 1;
  for (i=0; i<N; i++) {
    j = (rand()*1./RAND_MAX)*(N-1); // a random val in [0,N-1]
    swap(a, i, j);
  }
  return a;
}

// Bubble sort for the base cases
void bubbleSort(int *a, int low, int high) {
  if (low >= high) return;
  int i, j;
  for (i = low; i <= high; i++)
    for (j = i+1; j <= high; j++) 
      if (a[i] > a[j])
	swap(a, i, j);
}

// Partition an array range a[low, high] into two sub-ranges
int partition(int *a, int low, int high) {
  int pivot, middle, i;
  pivot = a[high]; 	// use highest element as pivot 
  middle = low;
  for (i=low; i<high; i++)
    if (a[i] < pivot) {
      swap(a, i, middle);
      middle++;
    }
  swap(a, high, middle);
  return middle;
}

// QuickSort an array range a[low, high]
void quickSort(int *a, int low, int high) {
  if (high - low < MINSIZE) {
    bubbleSort(a, low, high);
  } else {
    int middle = partition(a, low, high);
    if (low < middle)
      quickSort(a, low, middle-1);
    if (middle < high)
      quickSort(a, middle+1, high);
  }
}
 
// Print an array
void printArray(int *a, int N) {
  int i;
  for (i=0; i<N; i++)
    printf("%d, ", a[i]);
  printf("\n");
}

// Verify a sorted array
void verifyArray(int *a, int N) {
  int i;
  for (i=0; i<N-2; i++)
    if (a[i]>a[i+1]) {
      printf("FAILED: a[%d]=%d, a[%d]=%d\n", 
	     i, a[i], i+1, a[i+1]);
      return;
    }
  printf("Result verified!\n");
}

// Main routine for testing quickSort
// 
int main(int argc, char **argv)
{
  int *a, N, i, j;
  
  assert(argc > 1);
  N = atoi(argv[1]);
  assert(N > 1);

  a = createArray(N);
  printf("Initial array:\n");
  printArray(a, N);

  quickSort(a, 0, N-1);
  printf("After quickSort:\n");
  printArray(a, N);
  verifyArray(a, N);

}
